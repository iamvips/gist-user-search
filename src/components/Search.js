import React, { useState, useEffect } from "react";
import styled from "styled-components";
import Octicon from "react-octicon";
import { useSelector, useDispatch } from "react-redux";
import { getGistForUser } from "../services/gistService";
import "react-loader-spinner/dist/loader/css/react-spinner-loader.css";
import Loader from "react-loader-spinner";
import useDebounce from "../redux/hooks/useDebounce";
import useSearch from "../redux/hooks/useSearch";
import {
  searchUserGists,
  setUserGistResults,
  searchUserGistsFailure,
  emptySearchResults,
  setPageNumber,
} from "../redux/actions/searchAction";

const Search = () => {
  /** Default Debounce Time out 700ms*/
  const DEFAULT_TIMEOUT = 700;
  const [searchQuery, setSearchQuery] = useState("");
  const isFetching = useSelector((state) => state.searchReducer.isFetching);
  const dispatch = useDispatch();

  const debouncedSearchTerm = useDebounce(searchQuery, DEFAULT_TIMEOUT);
  /**
   * Hook to get the cached results from persistent store and
   * a handler to save the results for the new searches.
   */
  const { cachedResults, saveSearch } = useSearch();

  // Here's where the API call happens
  // We use useEffect since this is an asynchronous action
  useEffect(
    () => {
      /* Get the cahched Type */
      const cachedType = cachedResults["users"];

      /* Get the cached results based on the type*/
      const cached = cachedType && cachedType[debouncedSearchTerm];

      if (cached) {
        dispatch(setUserGistResults(cached));
        dispatch(setPageNumber(0));
      } else if (debouncedSearchTerm?.length >= 3) {
        // Make sure we have a value (user has entered something in input)
        // Fire off our API call
        handleSearchUserGistsRequest();
      } else if (debouncedSearchTerm?.length === 0) {
        dispatch(emptySearchResults());
      }
    },

    // This is the useEffect input array wherein
    // Our useEffect function will only execute if this value changes
    // and thanks to our hook it will only change if the original
    // value (searchTerm) hasn't changed for more than 700ms.
    // eslint-disable-next-line
    [debouncedSearchTerm]
  );

  const handleSearchUserGistsRequest = () => {
    dispatch(searchUserGists());

    getGistForUser(debouncedSearchTerm)
      .then((res) => {
        dispatch(setUserGistResults(res.data));
        saveSearch("USERS", debouncedSearchTerm, res.data);
      })
      .catch((err) => {
        dispatch(searchUserGistsFailure(err.message));
        dispatch(emptySearchResults());
      });
  };

  return (
    <Wrapper>
      <InputBox>
        <Octicon name="search" />
        <Input
          data-testid="search-input"
          placeholder="Search Gists for the username"
          onChange={(e) => setSearchQuery(e.target.value)}
        />

        {isFetching && (
          <Loader type="Puff" color="#00BFFF" height={25} width={25} />
        )}
      </InputBox>
    </Wrapper>
  );
};

const Wrapper = styled.div`
  padding: 8px;
  background-color: #ffffff;
  font-size: 14px;
  line-height: 1.5;
  border-radius: 6px;
  margin: 0 16px;
`;

const InputBox = styled.div`
  border-radius: 4px;
  display: flex;
  width: 400px;
`;

const Input = styled.input`
  border: none;
  width: 100%;
  font-size: 16px;

  &:focus {
    outline: 0;
  }
`;

export default Search;
