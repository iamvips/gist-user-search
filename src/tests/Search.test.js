import React from "react";
import renderer from "react-test-renderer";
import Search from "../components/Search";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

it("should render correctly", () => {
  const tree = renderer.create(<Search />).toJSON();
  expect(tree).toMatchSnapshot();
});
