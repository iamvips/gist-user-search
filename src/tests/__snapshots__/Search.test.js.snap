// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="sc-bdvvtL jcZpGJ"
>
  <div
    className="sc-gsDKAQ ldqIrX"
  >
    <span
      className="octicon octicon-search"
    />
    <input
      className="sc-dkPtRN lpkptG"
      data-testid="search-input"
      onChange={[Function]}
      placeholder="Search Gists for the username"
    />
  </div>
</div>
`;
